--- a/configure
+++ b/configure
@@ -1321,9 +1321,7 @@ fi
 
 for config in $mak_wilds; do
     target="$(basename "$config" .mak)"
-    if echo "$target_list_exclude" | grep -vq "$target"; then
-        default_target_list="${default_target_list} $target"
-    fi
+    target_list_exclude="${target_list_exclude},$target"
 done
 
 if test x"$show_help" = x"yes" ; then
@@ -1806,12 +1804,6 @@ if test "$tcg" = "enabled"; then
 fi
 
 if test -z "${target_list+xxx}" ; then
-    default_targets=yes
-    for target in $default_target_list; do
-        target_list="$target_list $target"
-    done
-    target_list="${target_list# }"
-else
     default_targets=no
     target_list=$(echo "$target_list" | sed -e 's/,/ /g')
     for target in $target_list; do
@@ -1821,7 +1813,6 @@ else
             *" $target "*)
                 ;;
             *)
-                error_exit "Unknown target name '$target'"
                 ;;
         esac
     done
